### ISSUE TITLE
[PriceLockup] - document use of legal copy

### ISSUE THREAD

**Issue by mike-bunce** [telus/tds-core#870]
## Description

- PriceLockup supports legals, but gives warnings as propTypes are set as strings for text fields. Legal support required for "price", "bottomText" and "topText". 

## Reproduction Steps

- Add legal tag to price field
- Content in component is wrapped in WithLegal component eg. `bottomText={<WithLegal content={bottomText} />}`
-  Page is rendered, legal renders too
- a propType warning appears in console.

## Reproduction Percentage (if applicable)

- 10/10 times

## Meta

- TDS version: v1.0.3 (current version)
- Willing to develop solution: Yes/No
- Has workaround: Yes
- High impact: No

**Comment by theetrain** [comment]
Hi @sketchidea, I'd like to gather more information:

1. When referring to Legal, are you referring to superscript numbers that link to legal content? If so, we can implement a feature similar to what's being done in [ExpandCollapse](https://github.com/telus/tds-core/blob/2b7eaf56c32e39b0526456013a7c45cd4b0acdf7/packages/ExpandCollapse/Panel/Panel.md)
2. I'm unfamiliar with the `WithLegal` component. Can that be contributed to tds-community?

**Comment by theetrain** [comment]
Here's a summary from a conversation I had with @sketchidea:

In brief:
- Enhance `WithLegal` to output as a string
- Enhance `PriceLockup` to accept superscript string templates

The `WithLegal` component can be enhanced by exporting a new method that returns superscript text as a string. The consumption of this method would resemble the following:

```jsx
import { withLegalText } from '@telusdigital/redux-contentful'

<PriceLockup price="100" bottomText={withLegalText(contentfulKey)} />
// withLegalText(contentfulKey) output == 'Limited time offer ^^1^^'
```

The double-caret output in `^^1^^` would then be processed by PriceLockup (not yet a feature) to wrap the `1` character with a superscript component. This string template behaviour is currently available in ExpandCollapse (see [panel documentation](https://github.com/telus/tds-core/blob/f3a833131bb860ac5e150f0967f34cae578fc046/packages/ExpandCollapse/Panel/Panel.md)).

@hamedmam do you see this as a viable enhancement?

**Comment by hamedmam** [comment]
@theetrain well, basically WithLegal is being used with many components inside Site-Builder. ex (Heading, Paragraph, and recently PriceLockup) the ExpandCollapse appears in cases where page has Legals and based on `legalId` it would show a description of that specific Legal. since the logic is bound to `redux-contentful` I am can't see how we might be able to natively support them in TDS but instead as per requested if PriceLockup's PropTypes can be updated in a way that they can also take `functions` would resolve the issue

**Comment by theetrain** [comment]
After a discussion with @hamedmam, @Nintendot, @sketchidea, @nikaukraine, and @simpleimpulse, we reached a proposal: 

Implement the `withLegalText()` method as a backwards-compatible enhancement.

## Why

Currently, many TDS components expect only strings to be passed as copy, which limits the potential of superscript or other rich-formatted copy. As a consequence, PropTypes warnings are thrown when superscript components are being used in TDS components that don't expect them.

There is a need to align on org-wide legal copy practices, convenient use of adding legal copy to pages using TDS, and for TDS components to be design-ready to support superscript text where applicable.

## What

The `withLegalText()` method, as described in my comment above, can be developed to return a string along with any necessary template strings such as double carets (`^^`). This would meet the needs of three audiences:

1. Content writers - within the CMS space, content writers can use rich formatting that would translate to branded components
2. Block consumers - blocks expecting legal copy can continue to make use of `<WithLegal />` so long as their PropTypes allow
3. TDS Consumers - applications that rely on TDS components such as `<PriceLockup />` can supply legal copy using the `withLegalText()` method that could still generate legal copy on its web page, while complying with the string propTypes expected by TDS components. TDS components would then necessarily parse string templates in order to properly render text

## How

- [ ] `withLegalText()` needs to be implemented to bridge compatibility between legal copy and TDS front end components. If this is not feasible, we can reassess this proposal
  - Owner: redux-contentful owners
- [ ] The TDS core team needs to audit all core components to find where superscript text should be accepted, and enhance them to support them
- [ ] Implement a sharable utility, tentatively named `@tds/util-string-parser` that makes the caret-parsing logic sharable across any component within or outside TDS. This is already being used by [ExpandCollapse](https://github.com/telus/tds-core/blob/cf705f416336d353bd2bdc23a4ada17d8b88074e/packages/ExpandCollapse/PanelWrapper/PanelWrapper.jsx#L23-L38)

**Comment by theetrain** [comment]
Conversation summary with @rbrander 

- Freeing up TDS components to use PropTypes.node would be an easy solution, but would require supplementary documentation to avoid unapproved patterns
- The `withLegalText()` method idea may not work since `<WithLegal />` works when it mounts (@jraff to follow up)
- It's important to have designers align on the freeform usage of legal or superscript text, and whether that can occur anywhere or only in certain places

**Comment by sean-saidmuradov** [comment]
I am observing some inconsistency in the way how a superscript is displayed on the page across the website, and want to share here the practice of displaying the superscripts based on different sources, including Wikipedia. Appreciate your comments or thoughts.  
(i) A superscript is a character (number, letter or symbol) that is set slightly above the normal lie of type. It is usually smaller than the rest of the text. To keep the visually consistent with the rest of the fonts, make them slightly heavier (medium or bold typography) than a reduced-size character would be. 
(ii) Superscript should be situated after all punctuation marks excluding a long dash  (—). 
(iii) o space between the superscript and punctuation mark, as a space pushes the superscript to the next line especially o a small or resized display.  
 (iv) Commas, if there are more then one superscript to be displayed, avoid a space between comma and next superscript. The reason is the same as above.

**Comment by sean-saidmuradov** [comment]
I am observing some inconsistency in the way how a superscript is displayed on the page across the website, and want to share here the practice of displaying the superscripts based on different sources, including Wikipedia. Appreciate your comments or thoughts.  
(i) A superscript is a character (number, letter or symbol) that is set slightly above the normal lie of type. It is usually smaller than the rest of the text. To keep the visually consistent with the rest of the fonts, make them slightly heavier (medium or bold typography) than a reduced-size character would be. 
(ii) Superscript should be situated after all punctuation marks excluding a long dash  (—). 
(iii) No space between the superscript and punctuation mark, as a space pushes the superscript to the next line especially on a small or resized display.  
 (iv) Commas, if there are more then one superscript to be displayed, avoid a space between comma and next superscript. The reason is the same as above.

**Comment by theetrain** [comment]
Here's an update. The Terms and Conditions component (#1048) will release soon. It will provide functional views for legal copy. We will keep this issue open to address Price Lockup separately.

**Comment by theetrain** [comment]
Update, this is being worked on now. You may follow this Jira story for updates: https://telusdigital.atlassian.net/browse/TDS-1353

**Comment by theetrain** [comment]
This feature is now available in [`@tds/core-price-lockup@2.1.0`](https://github.com/telus/tds-core/releases/tag/%40tds%2Fcore-price-lockup%402.1.0)
